_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,304 @@
+> It was fixed with by forcing the producer to send the consumer group metadata along with the offsets to commit ([KIP-447](https://cwiki.apache.org/confluence/display/KAFKA/KIP-447%3A+Producer+scalability+for+exactly+once+semantics)).
+
+To guarantee message ordering, a given producer can have at most one open transaction.
+Consumers with `isolation.level=read_committed` are not allowed to advance to offsets which include open transactions, while messages from aborted transactions are filtered out.
+
+The `TransactionCoordinator` is a module running inside every Kafka broker.
+Each coordinator owns a subset of the partitions in the internal `__transaction_state` topic (i.e. the partitions for which its broker is the leader).
+The coordinator automatically aborts any ongoing transaction that is not completed within `transaction.timeout.ms`.
+
+It is important to note that transactions are only supported inside a single Kafka cluster.
+If the transaction scope includes external systems, you would need to use an additional component such as `ChainedKafkaTransactionManager` from the SpringBoot project.
+This component is able to synchronize a database transaction with the Kafka transaction.
+
+#### Transaction states
+
+A transaction goes through the following states:
+
+1. **Undecided** (Ongoing)
+2. **Decided and unreplicated** (PrepareCommit\|PrepareAbort)
+3. **Decided and replicated** (CompleteCommit\|CompleteAbort)
+
+Non-transactional batches are considered decided immediately, but transactional batches are only decided when the corresponding commit or abort marker is written (control record).
+
+The high watermark (HW) is the offset of the last message that was successfully committed (replicated to all in-sync replicas).
+The first unstable offset (FUO) is the earliest offset that is part of the ongoing transaction, if present.
+The last stable offset (LSO) is the offset such that all lower offsets have been decided and it is always present.
+The LSO is equal to the FUO if it's lower than the HW, otherwise it's the HW (LSO <= HW <= LEO).
+
+![txn log](/assets/images/posts/2023-04-14-kafka-txn-log.png)
+
+#### Error handling
+
+If Kafka authorization is enabled, you need to configure the appropriate ACLs rules to avoid `TransactionalIdAuthorizationException`.
+
+In the following example we allow any `transactional.id` from any user.
+
+```sh
+$ kafka-acls.sh --bootstrap-server:9092 --command-config client.properties --add \
+  --allow-principal User:* --operation write --transactional-id *
+```
+
+If any of the send operations inside the transaction scope fails, the final `commitTransaction` will fail and throw the exception from the last failed send.
+When this happens, your application should call `abortTransaction` to reset the state and apply a retry strategy.
+Some errors cannot be recovered and the application should terminate (`ProducerFencedException`, `FencedInstanceIdException`, `OutOfOrderSequenceException`).
```

Comment:
Or close the current KafkaProducer instance and create a new one?

Reply:
Yeah, termination is too drastic if you are not running on K8s or need fast recovery times.